Util/model_parser_cxx/model_parser.H

```diff
@@ -202,6 +205,270 @@ Real interpolate_3d (const Real* loc, const Real* dx, int var_index, int nsub =
+                                                     amrex::min(model.state(i, model::idens) + drho,
+                                                                1.1_rt * model.state(i, model::idens)));
+                eos_state.rho = model.state(i, model::idens);
+
+                if (model.state(i, model::idens) < ambient::ambient_state[URHO]) {
+                    icutoff = i;
+                    fluff = true;
+                }
+
+                if (std::abs(drho) < model::hse_tol * model.state(i, model::idens)) {
+                    converged_hse = true;
+                    break;
+                }
+
+            }
+
+            if (!converged_hse && (!fluff)) {
+
+                std::cout << "Error: zone " <<  i << " did not converge in init_hse()" << std::endl;
+                std::cout << model.state(i, model::idens) << " " << model.state(i, model::itemp) << std::endl;
+                std::cout << p_want << " " << eos_state.p;
+                std::cout << drho << " " << model::hse_tol * model.state(i, model::idens);
+                amrex::Error("Error: HSE non-convergence.");
+
+            }
+
+            // Call the EOS to establish the final properties of this zone.
+
+            eos(eos_input_rt, eos_state);
+
+            model.state(i, model::ipres) = eos_state.p;
+
+            // Discretize the mass enclosed as (4 pi / 3) * rho * dr * (rl**2 + rl * rr + rr**2).
+
+            Real dM = (4.0_rt / 3.0_rt) * M_PI * model.state(i, model::idens) * dx *
+                      (rr * rr + rl * rr + rl * rl);
+            M_enclosed += dM;
+
+            if (i <= icutoff) {
+                // Also update the final WD mass if we're not in the ambient material.
+                mass += dM;
+            }
+
+        } // End loop over zones
+
+        radius = model.r(icutoff);
+
+        if (rho_c_old < 0.0_rt) {
+
+            // Not enough iterations yet -- use an arbitrary guess for the next iteration.
+
+            rho_c_old = rho_c;
+            rho_c = 0.5_rt * rho_c_old;
+
+        }
+        else {
+
+            // Check if we have converged.
+
+            if (std::abs(mass - mass_want) / mass_want < model::mass_tol) {
+                mass_converged = true;
+                break;
+            }
+
+            // Do a secant iteration:
+            // M_tot = M(rho_c) + dM/drho |_rho_c x drho + ...
+
+            drho_c = (mass_want - mass) / ((mass  - mass_old) / (rho_c - rho_c_old));
+
+            rho_c_old = rho_c;
+            rho_c = amrex::min(1.1e0_rt * rho_c_old, amrex::max((rho_c + drho_c), 0.9e0_rt * rho_c_old));
+
+        }
+
+        mass_old = mass;
+
+    } // End mass constraint loop
+
+    if (!mass_converged && max_mass_iter > 1) {
+        amrex::Error("ERROR: WD mass did not converge.");
+    }
+
+    central_density_want = model.state(0, model::idens);
+    mass_want = mass;
+
+    model::initialized = true;
+    model::npts = NPTS_MODEL;
```

Comment:
Note: I felt that if we were already specifying NPTS_MODEL at compile time that including an explicit npts argument to this routine is over-specifying the problem when we already have dx. The model points exist anyway, so we might as well just fill them all with fluff instead of arbitrarily cutting the model off after some number of points.

Reply:
I used to rely on knowing the actual number of points, to play some tricks in the BCs (I'd like to know where the actual edge of the model is).